File: src/lib/wordpress.ts
/**
 * WordPress integration.
 *
 * Wrappers for the WordPress backend commands (per-project site config) plus
 * the URL normalization used by the preview-pane setup gate.
 *
 * WordPress differs from the other theme integrations Ship Studio supports.
 * Shopify and HubSpot each ship a CLI (`shopify theme dev`, `hs cms theme
 * preview`) that renders *local* theme files against *remote* account data, so
 * their previews show local edits. WordPress has no such tool — it is PHP plus
 * a database, and runs only where both exist. So the preview reverse-proxies
 * the project's live site instead: real content, real rendering, but the
 * deployed theme rather than the working copy. Local edits appear only after
 * they are deployed.
 *
 * @module lib/wordpress
 */

import { invoke } from '@tauri-apps/api/core';

/** Get the live site a WordPress project previews (null = not connected). */
export async function getWordpressSiteUrl(projectPath: string): Promise<string | null> {
  return invoke<string | null>('get_wordpress_site_url', { projectPath });
}

/** Persist (or clear) the live site a WordPress project previews. */
export async function setWordpressSiteUrl(
  projectPath: string,
  siteUrl: string | null
): Promise<void> {
  return invoke<void>('set_wordpress_site_url', { projectPath, siteUrl });
}

/**
 * Probe a candidate site before saving it, so the setup gate can report a
 * dead host instead of handing the user a blank preview. Resolves to the HTTP
 * status, or null if the host was unreachable.
 */
export async function probeWordpressSite(siteUrl: string): Promise<number | null> {
  return invoke<number | null>('probe_wordpress_site', { siteUrl });
}

/**
 * Normalize whatever the user pastes into a bare `https://host` origin, or
 * null if it can't be one. Accepts `example.com`, `https://example.com/`,
 * `www.example.com/blog`, and hosts with an explicit port.
 *
 * Defaults to HTTPS when no scheme is given — a live WordPress site in 2026
 * is served over TLS, and guessing `http://` would trip the host's own
 * redirect on every request.
 */
export function normalizeSiteUrl(input: string): string | null {
  const trimmed = input.trim();
  if (!trimmed) return null;

  const withScheme = /^https?:\/\//i.test(trimmed) ? trimmed : `https://${trimmed}`;

  let parsed: URL;
  try {
    parsed = new URL(withScheme);
  } catch {
    return null;
  }
  if (parsed.protocol !== 'https:' && parsed.protocol !== 'http:') return null;
  if (!parsed.hostname || !parsed.hostname.includes('.')) {
    // Bare single labels ("localhost" aside) are almost always a typo.
    if (parsed.hostname !== 'localhost') return null;
  }
  // The backend rejects anything carrying a path — the proxy owns paths.
  return `${parsed.protocol}//${parsed.host}`;
}

/** Host portion of a normalized origin, for display and for the proxy target. */
export function siteHost(siteUrl: string): string {
  try {
    return new URL(siteUrl).hostname;
  } catch {
    return siteUrl;
  }
}

/** Whether a normalized origin is served over TLS. */
export function siteIsTls(siteUrl: string): boolean {
  return siteUrl.startsWith('https://');
}

/** Port the proxy should connect to, defaulting to the scheme's. */
export function sitePort(siteUrl: string): number {
  try {
    const parsed = new URL(siteUrl);
    if (parsed.port) return Number(parsed.port);
    return parsed.protocol === 'https:' ? 443 : 80;
  } catch {
    return 443;
  }
}

/** SSH connection details for running `wp` against a WordPress install. */
export interface WordpressSsh {
  host?: string | null;
  user?: string | null;
  keyPath?: string | null;
  wpPath?: string | null;
}

/** A complete connection: what to preview, and how to reach it. */
export interface WordpressConnection {
  siteUrl: string;
  ssh: WordpressSsh;
}

/**
 * Derive a full WP Engine connection from an install name. WP Engine names
 * everything after the install, so one input yields host, user, path and URL.
 */
export async function deriveWpEngineConfig(install: string): Promise<WordpressConnection> {
  return invoke<WordpressConnection>('derive_wpengine_config', { install });
}

/** Read the stored SSH connection for a project (null = not configured). */
export async function getWordpressSsh(projectPath: string): Promise<WordpressSsh | null> {
  return invoke<WordpressSsh | null>('get_wordpress_ssh', { projectPath });
}

/** Persist (or clear) the SSH connection for a project. */
export async function setWordpressSsh(
  projectPath: string,
  ssh: WordpressSsh | null
): Promise<void> {
  return invoke<void>('set_wordpress_ssh', { projectPath, ssh });
}

/**
 * Drop the pending marker once the project has real WordPress files, so
 * detection stands on its own evidence. Resolves true if it cleared one.
 */
export async function reconcileWordpressPending(projectPath: string): Promise<boolean> {
  return invoke<boolean>('reconcile_wordpress_pending', { projectPath });
}

/** Mark a project as WordPress before any theme files exist. */
export async function setWordpressPending(projectPath: string, pending: boolean): Promise<void> {
  return invoke<void>('set_wordpress_pending', { projectPath, pending });
}

/**
 * The `ssh … wp …` command line for this connection, for display and for
 * handing to the agent. Returns null unless the connection is complete.
 */
export function wpSshCommand(ssh: WordpressSsh | null): string | null {
  if (!ssh?.host || !ssh.user || !ssh.wpPath) return null;
  const key = ssh.keyPath ? `-i ${ssh.keyPath} ` : '';
  return `ssh ${key}${ssh.user}@${ssh.host} "wp <command> --path=${ssh.wpPath}"`;
}

/**
 * Prompt for the agent to stand up a local WordPress site.
 *
 * PHP + SQLite rather than Docker: it needs no daemon, no account and no
 * payment, so the agent can run the whole thing unattended. The resulting
 * site is an ordinary HTTP origin on localhost, which the preview proxy
 * already handles as a non-remote target — no extra wiring needed.
 */
export function localSiteSetupPrompt(projectPath: string, port: number): string {
  return [
    `Set up a local WordPress site for this project at ${projectPath}.`,
    '',
    'Steps:',
    '1. Ensure PHP and WP-CLI are available (`brew install php wp-cli` on macOS).',
    '2. `wp core download --path=wp` inside the project.',
    '3. Add the SQLite drop-in so no MySQL server is needed:',
    '   download the `sqlite-database-integration` plugin into',
    '   `wp/wp-content/plugins/`, then copy its `db.copy` to',
    '   `wp/wp-content/db.php` (edit the placeholder paths as its README says).',
    '4. `wp config create --dbname=wordpress --dbuser=root --skip-check --path=wp`',
    '5. `wp core install --url=http://localhost:' +
      port +
      ' --title="My Site" --admin_user=admin --admin_password=admin --admin_email=admin@example.com --path=wp`',
    '',
    'Notes:',
    "- WP-CLI's default 128M memory limit can kill `wp core download` during",
    '  extraction. If that happens, run commands as',
    '  `php -d memory_limit=512M $(which wp) <command> --path=wp`.',
    `- The install URL must be http://localhost:${port} exactly — WordPress`,
    '  stores it in the database and serves absolute links from it.',
    '- **Do not leave a server running.** Ship Studio starts and supervises it',
    "  as the project's dev server; a server backgrounded from your shell dies",
    '  with the session and the preview would break. To verify, start one',
    '  briefly, check it responds, then stop it.',
    '',
    'When the install is complete, tell me — Ship Studio will serve it.',
    'Do not modify anything outside this project.',
  ].join('\n');
}

/**
 * Prompt for the agent to set up SSH access to an existing install, so it can
 * run `wp` remotely. Content and config live in the site's database, not the
 * repo, so this is the only way an agent can change them.
 */
export function sshSetupPrompt(ssh: WordpressSsh): string {
  const keyPath = ssh.keyPath ?? '~/.ssh/wordpress_site';
  return [
    'Set up SSH access to this WordPress install so `wp` can be run remotely.',
    '',
    `1. If ${keyPath} does not exist, generate it:`,
    `   \`ssh-keygen -t ed25519 -f ${keyPath} -N ""\``,
    `2. Print the public key (${keyPath}.pub) and tell me to add it to the`,
    "   host's control panel — I have to do that part myself in the browser.",
    '3. Once I confirm, verify the connection end to end:',
    `   \`ssh -i ${keyPath} ${ssh.user ?? '<user>'}@${ssh.host ?? '<host>'} "wp option get home --path=${ssh.wpPath ?? '<path>'}"\``,
    '',
    'Report the verified home URL back to me.',
  ].join('\n');
}

/**
 * Command Ship Studio runs to serve a local WordPress site.
 *
 * The site must be owned by Ship Studio rather than left running by the agent:
 * a server the agent backgrounds dies with its shell session, so the preview
 * works once and then 502s with nothing listening. Running it as the project's
 * dev server means it's supervised, visible in the dev-server panel, and comes
 * back when the project is reopened.
 */
export function localServerCommand(port: number, wpDir = 'wp'): string {
  // PHP's built-in server (what `wp server` wraps) is single-threaded by
  // default, so a page plus its assets serialize through one process and the
  // preview crawls. Workers are set via the environment, and `env` is used
  // rather than shell syntax because the PTY layer splits the command on
  // whitespace and never invokes a shell.
  return `env PHP_CLI_SERVER_WORKERS=${LOCAL_SERVER_WORKERS} wp server --port=${port} --path=${wpDir}`;
}

/** Worker processes for the local PHP server. */
const LOCAL_SERVER_WORKERS = 4;

/** What the dev server should do for a WordPress project. */
export interface LocalServerPlan {
  serve: boolean;
  port: number | null;
  installDir: string | null;
  /** Why, for the log line — every non-serving case is a distinct situation. */
  reason: string;
}

/**
 * Decide whether Ship Studio should serve this WordPress project locally.
 *
 * Shared by the dev server's start and restart paths *on purpose*: they were
 * separate copies of this logic, they drifted, and the restart path shipped
 * without a WordPress branch at all — falling through to `npm run dev` in a
 * project with no package.json. One decision, two callers.
 */
export function planLocalServer(
  siteUrl: string | null | undefined,
  installDir: string | null
): LocalServerPlan {
  const nothing = (reason: string): LocalServerPlan => ({
    serve: false,
    port: null,
    installDir: null,
    reason,
  });

  if (!siteUrl) return nothing('no site connected yet');
  if (!isLocalSite(siteUrl)) return nothing('project previews a live site; nothing to serve');
  if (!installDir) return nothing('no WordPress install on disk yet');

  // The site's URL is baked into its database at install time, so it has to be
  // served on that exact port — not a reserved dev-server port.
  return {
    serve: true,
    port: sitePort(siteUrl),
    installDir,
    reason: 'serving local WordPress install',
  };
}

/**
 * Find a WordPress install inside the project (returns its directory relative
 * to the project, `.` for the root, or null if there isn't one).
 *
 * This is how the setup flow knows a local site exists. It deliberately does
 * not depend on the site answering over HTTP: Ship Studio serves the site, and
 * only does so once one is connected, so probing first would deadlock.
 */
export async function detectLocalWordpress(projectPath: string): Promise<string | null> {
  return invoke<string | null>('detect_local_wordpress', { projectPath });
}

/** Whether a connected site is a local one Ship Studio should serve itself. */
export function isLocalSite(siteUrl: string | null | undefined): boolean {
  if (!siteUrl) return false;
  const host = siteHost(siteUrl);
  return host === 'localhost' || host === '127.0.0.1';
}

File: src-tauri/src/commands/wordpress.rs
//! # WordPress Integration Commands
//!
//! Per-project configuration for WordPress theme projects.
//!
//! Unlike Shopify and HubSpot, WordPress has no vendor CLI that renders a
//! local theme against remote content, so there is no dev server to spawn.
//! The preview instead reverse-proxies the project's live site (see
//! `proxy::start_preview_proxy`'s remote-host mode), which means the pane
//! shows *deployed* content — local theme edits are not reflected.
//!
//! These commands only answer "which site does this project preview?".

use crate::errors::CommandError;
use crate::types::{ProjectMetadata, WordpressSsh};
use crate::utils::validate_project_path;

/// Validate a site origin: `https://host` or `http://host`, no path, no
/// credentials, no whitespace. The value is interpolated into the proxy's
/// upstream connection and `Host` header, so it must not carry surprises.
fn validate_site_url(url: &str) -> Result<(), CommandError> {
    let rest = url
        .strip_prefix("https://")
        .or_else(|| url.strip_prefix("http://"))
        .ok_or_else(|| CommandError::Validation {
            field: "site_url".to_string(),
            reason: "must start with https:// or http://".to_string(),
        })?;

    if rest.is_empty() {
        return Err(CommandError::Validation {
            field: "site_url".to_string(),
            reason: "is missing a hostname".to_string(),
        });
    }
    if rest.contains('/') {
        return Err(CommandError::Validation {
            field: "site_url".to_string(),
            reason: "must be just the domain, with no path (e.g. https://example.com)".to_string(),
        });
    }
    if rest.contains('@') {
        return Err(CommandError::Validation {
            field: "site_url".to_string(),
            reason: "must not contain credentials".to_string(),
        });
    }
    // Hostname, optionally :port. Keep the charset tight — this string reaches
    // a TCP connect, a TLS SNI name and an HTTP header value.
    let host = rest.split(':').next().unwrap_or(rest);
    if host.is_empty()
        || !host
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b == b'-' || b == b'.')
    {
        return Err(CommandError::Validation {
            field: "site_url".to_string(),
            reason: format!("\"{host}\" is not a valid hostname"),
        });
    }
    if let Some(port) = rest.split(':').nth(1) {
        if port.parse::<u16>().is_err() {
            return Err(CommandError::Validation {
                field: "site_url".to_string(),
                reason: format!("\"{port}\" is not a valid port"),
            });
        }
    }
    Ok(())
}

fn read_metadata(metadata_path: &std::path::Path) -> ProjectMetadata {
    std::fs::read_to_string(metadata_path)
        .ok()
        .and_then(|contents| serde_json::from_str::<ProjectMetadata>(&contents).ok())
        .unwrap_or_default()
}

/// Gets the live site this WordPress project previews, or None if the user
/// hasn't connected one yet.
#[tauri::command]
#[tracing::instrument(fields(project = %project_path))]
pub async fn get_wordpress_site_url(project_path: String) -> Result<Option<String>, CommandError> {
    let project = validate_project_path(&project_path)?;
    let metadata_path = project.join(".shipstudio").join("project.json");

    if !metadata_path.exists() {
        return Ok(None);
    }

    Ok(read_metadata(&metadata_path).wordpress_site_url)
}

/// Sets (or clears, with None) the live site a WordPress project previews.
/// Expects a bare origin like `https://example.com` — the frontend normalizes
/// user input (trims, adds the scheme, strips the path) before calling.
#[tauri::command]
#[tracing::instrument(fields(project = %project_path))]
pub async fn set_wordpress_site_url(
    project_path: String,
    site_url: Option<String>,
) -> Result<(), CommandError> {
    if let Some(ref url) = site_url {
        validate_site_url(url)?;
    }

    let project = validate_project_path(&project_path)?;
    let shipstudio_dir = project.join(".shipstudio");
    let metadata_path = shipstudio_dir.join("project.json");

    let mut metadata = if metadata_path.exists() {
        read_metadata(&metadata_path)
    } else {
        ProjectMetadata::default()
    };

    metadata.wordpress_site_url = site_url;

    if !shipstudio_dir.exists() {
        std::fs::create_dir_all(&shipstudio_dir)
            .map_err(|e| format!("Failed to create .shipstudio directory: {e}"))?;
    }

    let contents = serde_json::to_string_pretty(&metadata)
        .map_err(|e| format!("Failed to serialize project metadata: {e}"))?;
    std::fs::write(&metadata_path, contents)
        .map_err(|e| format!("Failed to write project metadata: {e}"))?;

    Ok(())
}

/// Probe a candidate WordPress site before saving it, so the setup gate can
/// say "that host didn't respond" instead of handing the user a blank preview.
///
/// Sends a browser-like `User-Agent`: the origin may sit behind a CDN with bot
/// protection (WP Engine fronts sites with Cloudflare) that challenges default
/// client agents. Returns the final status code, or None if the host was
/// unreachable.
#[tauri::command]
#[tracing::instrument]
pub async fn probe_wordpress_site(site_url: String) -> Option<u16> {
    validate_site_url(&site_url).ok()?;

    let client = reqwest::Client::builder()
        .timeout(std::time::Duration::from_secs(15))
        .user_agent(
            "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 \
             (KHTML, like Gecko) Chrome/131.0.0.0 Safari/537.36",
        )
        .build()
        .ok()?;

    client
        .get(&site_url)
        .send()
        .await
        .ok()
        .map(|resp| resp.status().as_u16())
}

/// Derive a full WP Engine connection from an install name.
///
/// WP Engine names everything after the install, so one input yields the SSH
/// host, user, WordPress path and public URL:
/// `myinstall` → `myinstall@myinstall.ssh.wpengine.net`,
/// `--path=/sites/myinstall`, `https://myinstall.wpenginepowered.com`.
#[tauri::command]
#[tracing::instrument]
pub fn derive_wpengine_config(install: String) -> Result<WordpressConnection, CommandError> {
    let install = install.trim().to_ascii_lowercase();
    if install.is_empty()
        || !install
            .bytes()
            .all(|b| b.is_ascii_alphanumeric() || b == b'-')
    {
        return Err(CommandError::Validation {
            field: "install".to_string(),
            reason: "WP Engine install names are letters, numbers and hyphens".to_string(),
        });
    }
    Ok(WordpressConnection {
        site_url: format!("https://{install}.wpenginepowered.com"),
        ssh: WordpressSsh {
            host: Some(format!("{install}.ssh.wpengine.net")),
            user: Some(install.clone()),
            key_path: Some(format!("~/.ssh/{install}_wpengine")),
            wp_path: Some(format!("/sites/{install}")),
        },
    })
}

/// A complete WordPress connection: what to preview, and how to reach it.
#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct WordpressConnection {
    pub site_url: String,
    pub ssh: WordpressSsh,
}

/// Read the stored SSH connection for a project, if one is configured.
#[tauri::command]
#[tracing::instrument(fields(project = %project_path))]
pub async fn get_wordpress_ssh(project_path: String) -> Result<Option<WordpressSsh>, CommandError> {
    let project = validate_project_path(&project_path)?;
    let metadata_path = project.join(".shipstudio").join("project.json");
    if !metadata_path.exists() {
        return Ok(None);
    }
    Ok(read_metadata(&metadata_path).wordpress_ssh)
}

/// Persist (or clear) the SSH connection for a project.
#[tauri::command]
#[tracing::instrument(fields(project = %project_path))]
pub async fn set_wordpress_ssh(
    project_path: String,
    ssh: Option<WordpressSsh>,
) -> Result<(), CommandError> {
    let project = validate_project_path(&project_path)?;
    let shipstudio_dir = project.join(".shipstudio");
    let metadata_path = shipstudio_dir.join("project.json");

    let mut metadata = if metadata_path.exists() {
        read_metadata(&metadata_path)
    } else {
        ProjectMetadata::default()
    };
    metadata.wordpress_ssh = ssh;
    write_metadata(&shipstudio_dir, &metadata_path, &metadata)
}

/// Mark a project as WordPress before any theme files exist, so a
/// freshly-created project opens into the WordPress setup flow instead of
/// detecting as `unknown` and showing no preview at all.
#[tauri::command]
#[tracing::instrument(fields(project = %project_path))]
pub async fn set_wordpress_pending(
    project_path: String,
    pending: bool,
) -> Result<(), CommandError> {
    let project = validate_project_path(&project_path)?;
    let shipstudio_dir = project.join(".shipstudio");
    let metadata_path = shipstudio_dir.join("project.json");

    let mut metadata = if metadata_path.exists() {
        read_metadata(&metadata_path)
    } else {
        ProjectMetadata::default()
    };
    metadata.wordpress_pending = if pending { Some(true) } else { None };
    write_metadata(&shipstudio_dir, &metadata_path, &metadata)
}

/// Drop the `wordpress_pending` marker once the project carries real
/// WordPress files, so detection stands on its own evidence.
///
/// The marker must NOT be cleared merely because a site was connected: a
/// project that previews a live site legitimately has no WordPress files of
/// its own, and clearing it there would make the project detect as `unknown`
/// and lose its preview entirely.
#[tauri::command]
#[tracing::instrument(fields(project = %project_path))]
pub async fn reconcile_wordpress_pending(project_path: String) -> Result<bool, CommandError> {
    let project = validate_project_path(&project_path)?;
    let metadata_path = project.join(".shipstudio").join("project.json");
    if !metadata_path.exists() {
        return Ok(false);
    }
    let mut metadata = read_metadata(&metadata_path);
    if metadata.wordpress_pending != Some(true) {
        return Ok(false);
    }
    // Real files = a theme, or a full install.
    let has_files = crate::commands::projects::detection::is_wordpress_project(&project)
        || find_wordpress_install(&project).is_some();
    if !has_files {
        return Ok(false);
    }
    metadata.wordpress_pending = None;
    write_metadata(&project.join(".shipstudio"), &metadata_path, &metadata)?;
    Ok(true)
}

fn write_metadata(
    shipstudio_dir: &std::path::Path,
    metadata_path: &std::path::Path,
    metadata: &ProjectMetadata,
) -> Result<(), CommandError> {
    if !shipstudio_dir.exists() {
        std::fs::create_dir_all(shipstudio_dir)
            .map_err(|e| format!("Failed to create .shipstudio directory: {e}"))?;
    }
    let contents = serde_json::to_string_pretty(metadata)
        .map_err(|e| format!("Failed to serialize project metadata: {e}"))?;
    std::fs::write(metadata_path, contents)
        .map_err(|e| format!("Failed to write project metadata: {e}"))?;
    Ok(())
}

/// Find a WordPress *install* (core files, not just a theme) inside the
/// project — the project root or a direct child such as `wp/`.
///
/// This is what proves a locally-provisioned site exists. The setup flow can't
/// rely on the site answering over HTTP, because Ship Studio is the thing that
/// serves it and it won't do that until a site is connected: probing first
/// would deadlock. Files on disk break the cycle.
#[tauri::command]
#[tracing::instrument(fields(project = %project_path))]
pub async fn detect_local_wordpress(project_path: String) -> Result<Option<String>, CommandError> {
    let project = validate_project_path(&project_path)?;
    Ok(find_wordpress_install(&project))
}

/// Core files (`wp-config.php` + `wp-includes/`) mark a WordPress install, as
/// opposed to a theme-only project. Checks the root and each direct child.
fn find_wordpress_install(project: &std::path::Path) -> Option<String> {
    let is_install = |dir: &std::path::Path| {
        dir.join("wp-config.php").exists() && dir.join("wp-includes").is_dir()
    };
    if is_install(project) {
        return Some(".".to_string());
    }
    let entries = std::fs::read_dir(project).ok()?;
    let mut names: Vec<String> = entries
        .flatten()
        .filter(|e| e.path().is_dir())
        .filter_map(|e| e.file_name().to_str().map(str::to_owned))
        .filter(|n| !n.starts_with('.') && n != "node_modules")
        .collect();
    names.sort();
    names.into_iter().find(|n| is_install(&project.join(n)))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn accepts_plain_origins() {
        assert!(validate_site_url("https://example.com").is_ok());
        assert!(validate_site_url("http://example.com").is_ok());
        assert!(validate_site_url("https://myinstall.wpenginepowered.com").is_ok());
        assert!(validate_site_url("http://localhost:8888").is_ok());
    }

    #[test]
    fn rejects_paths_and_missing_scheme() {
        assert!(validate_site_url("example.com").is_err());
        assert!(validate_site_url("https://example.com/blog").is_err());
        // A trailing slash is still a path — the frontend strips it first.
        assert!(validate_site_url("https://example.com/").is_err());
        assert!(validate_site_url("https://").is_err());
    }

    #[test]
    fn finds_a_wordpress_install_on_disk() {
        let tmp = std::env::temp_dir().join(format!("ss-wp-install-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&tmp);
        std::fs::create_dir_all(tmp.join("wp").join("wp-includes")).unwrap();
        std::fs::write(tmp.join("wp").join("wp-config.php"), "<?php").unwrap();
        assert_eq!(find_wordpress_install(&tmp).as_deref(), Some("wp"));

        // An install at the project root, not a subdirectory.
        let root = std::env::temp_dir().join(format!("ss-wp-root-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&root);
        std::fs::create_dir_all(root.join("wp-includes")).unwrap();
        std::fs::write(root.join("wp-config.php"), "<?php").unwrap();
        assert_eq!(find_wordpress_install(&root).as_deref(), Some("."));

        // A theme-only project is not an install — core files are the marker.
        let bare = std::env::temp_dir().join(format!("ss-wp-bare-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&bare);
        std::fs::create_dir_all(&bare).unwrap();
        assert_eq!(find_wordpress_install(&bare), None);

        let _ = std::fs::remove_dir_all(&tmp);
        let _ = std::fs::remove_dir_all(&root);
        let _ = std::fs::remove_dir_all(&bare);
    }

    #[test]
    fn pending_marker_only_clears_once_real_files_exist() {
        // The marker is what keeps a live-site project (which legitimately has
        // no WordPress files of its own) detected as WordPress. Clearing it on
        // "site connected" would make such a project detect as `unknown` and
        // silently lose its preview — so presence of files is the only trigger.
        let bare = std::env::temp_dir().join(format!("ss-wp-pend-{}", std::process::id()));
        let _ = std::fs::remove_dir_all(&bare);
        std::fs::create_dir_all(&bare).unwrap();
        assert!(find_wordpress_install(&bare).is_none());
        assert!(!crate::commands::projects::detection::is_wordpress_project(
            &bare
        ));

        // Give it a real install: now detection can stand on its own.
        std::fs::create_dir_all(bare.join("wp").join("wp-includes")).unwrap();
        std::fs::write(bare.join("wp").join("wp-config.php"), "<?php").unwrap();
        assert!(find_wordpress_install(&bare).is_some());

        let _ = std::fs::remove_dir_all(&bare);
    }

    #[test]
    fn derives_wpengine_connection_from_install_name() {
        // The shape WP Engine uses: every field derives from the install name.
        let c = derive_wpengine_config("myinstall".to_string()).unwrap();
        assert_eq!(c.site_url, "https://myinstall.wpenginepowered.com");
        assert_eq!(
            c.ssh.host.as_deref(),
            Some("myinstall.ssh.wpengine.net")
        );
        assert_eq!(c.ssh.user.as_deref(), Some("myinstall"));
        assert_eq!(c.ssh.wp_path.as_deref(), Some("/sites/myinstall"));
        assert_eq!(
            c.ssh.key_path.as_deref(),
            Some("~/.ssh/myinstall_wpengine")
        );
    }

    #[test]
    fn rejects_bad_install_names() {
        assert!(derive_wpengine_config("".to_string()).is_err());
        assert!(derive_wpengine_config("has space".to_string()).is_err());
        // Command-injection shaped input must never reach an ssh invocation.
        assert!(derive_wpengine_config("a; rm -rf /".to_string()).is_err());
        assert!(derive_wpengine_config("a/../b".to_string()).is_err());
    }

    #[test]
    fn rejects_injection_shaped_input() {
        assert!(validate_site_url("https://user:pass@example.com").is_err());
        assert!(validate_site_url("https://exa mple.com").is_err());
        assert!(validate_site_url("https://example.com:notaport").is_err());
        assert!(validate_site_url("https://example.com\r\nX-Evil: 1").is_err());
    }
}
